File: cloudtimes-admin/src/main/java/com/cloudtimes/web/controller/store/CtStoreController.java
package com.cloudtimes.web.controller.store;

import java.util.List;
import javax.servlet.http.HttpServletResponse;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import com.cloudtimes.common.annotation.Log;
import com.cloudtimes.common.core.controller.BaseController;
import com.cloudtimes.common.core.domain.AjaxResult;
import com.cloudtimes.common.enums.BusinessType;
import com.cloudtimes.store.domain.CtStore;
import com.cloudtimes.store.service.ICtStoreService;
import com.cloudtimes.common.utils.poi.ExcelUtil;
import com.cloudtimes.common.core.page.TableDataInfo;

/**
 * 门店Controller
 * 
 * @author wangxp
 * @date 2023-01-13
 */
@RestController
@RequestMapping("/store/store")
public class CtStoreController extends BaseController
{
    @Autowired
    private ICtStoreService ctStoreService;

    /**
     * 查询门店列表
     */
    @PreAuthorize("@ss.hasPermi('store:store:list')")
    @GetMapping("/list")
    public TableDataInfo list(CtStore ctStore)
    {
        startPage();
        List<CtStore> list = ctStoreService.selectCtStoreList(ctStore);
        return getDataTable(list);
    }

    /**
     * 导出门店列表
     */
    @PreAuthorize("@ss.hasPermi('store:store:export')")
    @Log(title = "门店", businessType = BusinessType.EXPORT)
    @PostMapping("/export")
    public void export(HttpServletResponse response, CtStore ctStore)
    {
        List<CtStore> list = ctStoreService.selectCtStoreList(ctStore);
        ExcelUtil<CtStore> util = new ExcelUtil<CtStore>(CtStore.class);
        util.exportExcel(response, list, "门店数据");
    }

    /**
     * 获取门店详细信息
     */
    @PreAuthorize("@ss.hasPermi('store:store:query')")
    @GetMapping(value = "/{id}")
    public AjaxResult getInfo(@PathVariable("id") Long id)
    {
        return AjaxResult.success(ctStoreService.selectCtStoreById(id));
    }

    /**
     * 新增门店
     */
    @PreAuthorize("@ss.hasPermi('store:store:add')")
    @Log(title = "门店", businessType = BusinessType.INSERT)
    @PostMapping
    public AjaxResult add(@RequestBody CtStore ctStore)
    {
        return toAjax(ctStoreService.insertCtStore(ctStore));
    }

    /**
     * 修改门店
     */
    @PreAuthorize("@ss.hasPermi('store:store:edit')")
    @Log(title = "门店", businessType = BusinessType.UPDATE)
    @PutMapping
    public AjaxResult edit(@RequestBody CtStore ctStore)
    {
        return toAjax(ctStoreService.updateCtStore(ctStore));
    }

    /**
     * 删除门店
     */
    @PreAuthorize("@ss.hasPermi('store:store:remove')")
    @Log(title = "门店", businessType = BusinessType.DELETE)
	@DeleteMapping("/{ids}")
    public AjaxResult remove(@PathVariable Long[] ids)
    {
        return toAjax(ctStoreService.deleteCtStoreByIds(ids));
    }
}

File: cloudtimes-business/src/main/java/com/cloudtimes/app/controller/agent/CtStoreController.kt
package com.cloudtimes.app.controller.agent

import com.cloudtimes.account.dto.request.QueryByUserIdRequest
import com.cloudtimes.agent.dto.request.AgentStoreListRequest
import com.cloudtimes.account.dto.response.StoreAndCommission
import com.cloudtimes.agent.dto.request.AgentStoreDetailRequest
import com.cloudtimes.agent.dto.request.StoreProfitRequest
import com.cloudtimes.agent.dto.response.AgentStoreOnlineStats
import com.cloudtimes.agent.dto.response.AgentStoreProfitStats
import com.cloudtimes.agent.dto.response.StoreOrderDetail
import com.cloudtimes.agent.dto.response.OrderMonthlyStats
import com.cloudtimes.agent.service.ICtUserAgentService
import com.cloudtimes.app.constant.PrefixPathConstants
import com.cloudtimes.app.controller.system.SmsController
import com.cloudtimes.common.core.controller.BaseController
import com.cloudtimes.common.core.domain.AjaxResult
import com.cloudtimes.common.core.domain.RestPageResult
import com.cloudtimes.common.core.domain.RestResult
import com.cloudtimes.hardwaredevice.dto.request.QueryOrdersByDate
import com.cloudtimes.hardwaredevice.dto.request.RegisterStoreRequest
import com.cloudtimes.hardwaredevice.dto.request.QueryOrdersByMonth
import com.cloudtimes.hardwaredevice.service.ICtStoreService
import com.cloudtimes.supervise.service.ICtOrderService
import io.swagger.annotations.Api
import io.swagger.annotations.ApiOperation
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import javax.validation.Valid


// 泛型具体化
class StoreAndCommissionPage() : RestPageResult<StoreAndCommission>()
class StoreAndCommissionDetail() : RestResult<StoreAndCommission>()
class AgentShopStatsPage() : RestPageResult<AgentStoreOnlineStats>()
class MonthlyOrderResponse() : RestPageResult<OrderMonthlyStats>()
class DailyOrderResponse() : RestPageResult<StoreOrderDetail>()
class StoreProfitResponse() : RestResult<AgentStoreProfitStats>()

/**
 * 用户Controller
 *
 * @author 沈兵
 * @date 2023-01-17
 */
@RestController
@RequestMapping(PrefixPathConstants.WX_OFFICIAL_PATH_PREFIX + "/store")
@Api(tags = ["代理-门店"])
class CtStoreController : BaseController() {
    @Autowired
    private lateinit var ctUserAgentService: ICtUserAgentService

    @Autowired
    private lateinit var ctOrderService: ICtOrderService

    @Autowired
    private lateinit var ctStoreService: ICtStoreService

    @Autowired
    private lateinit var smsController: SmsController

    /**
     * 查询代理门店
     */
    @ApiOperation("查询代理门店列表")
    @PostMapping("/list")
    fun list(@Valid @RequestBody request: AgentStoreListRequest): StoreAndCommissionPage {
        startPage(request.pageNum, request.pageSize)
        val storeAndCommissionList = ctUserAgentService.selectCtAgentShopList(request)
        val pageData = getDataTable(storeAndCommissionList)
        return StoreAndCommissionPage().apply {
            total = pageData.total
            data = storeAndCommissionList
        }
    }

    /**
     * 查询代理门店详情
     */
    @ApiOperation("查询代理门店详情")
    @PostMapping("/detail")
    fun detail(@Valid @RequestBody request: AgentStoreDetailRequest): StoreAndCommissionDetail {
        val detail = ctUserAgentService.selectCtAgentShopDetail(request)
        return StoreAndCommissionDetail().apply {
            data = detail
        }
    }

    /**
     * 查询代理门店没有订单统计
     */
    @ApiOperation("查询代理门店每月订单统计")
    @PostMapping("/monthly_order_stats")
    fun monthlyOrderStats(@Valid @RequestBody request: QueryOrdersByMonth): MonthlyOrderResponse {
        val list = ctOrderService.selectMonthlyOrderStats(request)

        return MonthlyOrderResponse().apply {
            data = list
            total = list.size.toLong()
        }
    }


    /**
     * 查询代理门店没有订单
     */
    @ApiOperation("查询代理门店每日订单")
    @PostMapping("/daily_orders")
    fun dailyOrder(@Valid @RequestBody request: QueryOrdersByDate): DailyOrderResponse {
        startPage(request.pageNum, request.pageSize)

        val list = ctOrderService.selectDailyOrders(request)
        val page = getDataTable(list)

        return DailyOrderResponse().apply {
            data = list
            total = page.total
        }
    }

    /**
     * 查询代理门店上线统计
     */
    @ApiOperation("查询代理门店上线统计")
    @PostMapping("/online_stats")
    fun onlineStats(@Valid @RequestBody request: QueryByUserIdRequest): AgentShopStatsPage {
        val list = ctUserAgentService.selectCtAgentShopOnlineStats(request.userId)
        return AgentShopStatsPage().apply {
            data = list
            total = list.size.toLong()
        }
    }

    /**
     * 查询代理门店收益统计
     */
    @ApiOperation("查询代理门店收益统计")
    @PostMapping("/profit_stats")
    fun profitStats(@Valid @RequestBody request: StoreProfitRequest): StoreProfitResponse {
        val profits = ctUserAgentService.selectCtAgentShopProfitStats(request)
        return StoreProfitResponse().apply {
            data = profits
        }
    }

    @ApiOperation("门店注册")
    @PostMapping("/register")
    fun register(@Valid @RequestBody request: RegisterStoreRequest): AjaxResult {
        // Step 1. 校验手机验证码
        smsController.validateSMS(request.mobile, request.verifyCode, request.verifyUUID)

        // Step 2. 业务处理
        ctStoreService.registerStore(request)
        return AjaxResult.success()
    }

}
